program/src/processor.rs

```diff
@@ -601,83 +650,138 @@ pub fn process_withdraw_inactive_stake(
+            accounts: &accounts,
+            clock: &clock,
+            stake_history: &stake_history,
+            stake_account: provided_stake_account,
+        };
+
+        let stake_account = get_stake_account(&withdraw_opts, seed)?;
+        let amount = (stake_account.balance.inactive
+            - Lamports(rent.minimum_balance(provided_stake_account.data_len())))
+        .expect("Should have at least the payed rent");
 
-        excess_removed = (excess_removed + excess_removed_here)?;
-        observed_total = (observed_total + account_balance)?;
+        withdraw_inactive_sol(&withdraw_opts, seed, lido.stake_authority_bump_seed, amount)?;
+
+        excess_removed = (excess_removed + amount)?;
+        stake_observed_total = (stake_observed_total + account_balance)?;
     }
 
-    if observed_total < validator.entry.stake_accounts_balance {
+    if stake_observed_total < validator.entry.effective_stake_balance() {
         // Solana has no slashing at the time of writing, and only Solido can
         // withdraw from these accounts, so we should not observe a decrease in
         // balance.
         msg!(
             "Observed balance of {} is less than tracked balance of {}.",
-            observed_total,
-            validator.entry.stake_accounts_balance
+            stake_observed_total,
+            validator.entry.effective_stake_balance()
         );
         msg!("This should not happen, aborting ...");
         return Err(LidoError::ValidatorBalanceDecreased.into());
     }
 
     // We tracked in `stake_accounts_balance` what we put in there ourselves, so
     // the excess is a donation by some joker.
-    let donation = (observed_total - validator.entry.stake_accounts_balance)
+    let donation = (stake_observed_total - validator.entry.effective_stake_balance())
         .expect("Does not underflow because observed_total >= stake_accounts_balance.");
     msg!("{} in donations observed.", donation);
 
+    // Try to withdraw from unstake accounts.
+    let mut unstake_removed = Lamports(0);
+    let mut unstake_observed_total = Lamports(0);
+    for (seed, unstake_account) in validator
+        .entry
+        .unstake_seeds
+        .into_iter()
+        .zip(unstake_accounts)
+    {
+        let (unstake_account_address, _bump_seed) =
+            validator.find_unstake_account_address(program_id, accounts.lido.key, seed);
+
+        let account_balance = check_address_and_get_balance(
+            &unstake_account_address,
+            unstake_account,
+            seed,
+            StakeType::Unstake,
+        )?;
+
+        let withdraw_opts = WithdrawExcessOpts {
+            accounts: &accounts,
+            clock: &clock,
+            stake_history: &stake_history,
+            stake_account: unstake_account,
+        };
+        let stake_account = get_stake_account(&withdraw_opts, seed)?;
+
+        // If validator's stake is at the beginning, try to withdraw.
```

Comment:
It would be good to also explain why we don’t withdraw if it’s not at the beginning.